aten/src/ATen/native/Blas.cpp

```diff
@@ -69,31 +44,59 @@ Tensor &addmv_out(const Tensor &self, const Tensor &mat, const Tensor &vec, cons
     }
-  } else {
-    if (!result.is_same(self_)) {
-      at::native::copy_(result, self_);
-    }
-    if (result.numel() != 0) {
-      at::_addmv_impl_(result, self_, mat, vec, beta, alpha);
-    }
   }
+}
+
 
-  } // scope of NoNamesGuard
-  at::namedinference::propagate_names_for_addmv(result, mat, vec, self);
-  return result;
 }
 
-Tensor addmv(const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha) {
-  Tensor result = at::empty({mat.size(0)}, mat.options());
-  return native::addmv_out(self, mat, vec, beta, alpha, result);
+
+namespace native {
+
+template<typename scalar_t>
+void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t lda, scalar_t *x, int64_t incx, scalar_t beta, scalar_t *y, int64_t incy);
+
+template<typename scalar_t>
+scalar_t dot_impl(int64_t n, scalar_t *x, int64_t incx, scalar_t *y, int64_t incy);
+
+template<typename scalar_t>
+scalar_t vdot_impl(int64_t n, scalar_t *x, int64_t incx, scalar_t *y, int64_t incy);
+
+constexpr inline bool lda_cond(int64_t m, int64_t n, int64_t lda) {
+  return n == 1 || lda >= std::max<int64_t>(1L, m);
 }
 
-Tensor &addmv_(Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha) {
-  return native::addmv_out(self, mat, vec, beta, alpha, self);
+
+
+
+TORCH_IMPL_FUNC(addmv_out_cpu)(const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta_, const Scalar& alpha_, const Tensor& result) {
+// empty matrix was handled in meta function
+  if (mat.numel() != 0) {
+    if (result.numel() != 0) {
+      auto r_stride = result.stride(0);
+      AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(kBFloat16, mat.scalar_type(), "addmv_impl_cpu", [&] {
+        auto beta = beta_.to<scalar_t>();
+        auto alpha = alpha_.to<scalar_t>();
+        if (mat.stride(0) == 1 && lda_cond(mat.size(0), mat.size(1), mat.stride(1))) {
+          gemv<scalar_t>('n', mat.size(0), mat.size(1), alpha, mat.data_ptr<scalar_t>(), mat.stride(1),
+              vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
+        }
+        else if (mat.stride(1) == 1 && lda_cond(mat.size(1), mat.size(0), mat.stride(0))) {
+          gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, mat.data_ptr<scalar_t>(), mat.stride(0),
+              vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
+        }
+        else {
+          Tensor cmat = mat.contiguous();
+          gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, cmat.data_ptr<scalar_t>(), cmat.stride(0),
+              vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
+        }
+      });
+    }
+  }
 }
 
 Tensor &mv_out(const Tensor &self, const Tensor &vec, Tensor& result) {
-  return native::addmv_out(result, self, vec, 0, 1, result);
+  return at::addmv_out(result, result, self, vec, 0, 1);
```

Comment:
this adds another dispatch, but I don't see how it can be avoided.

Reply:
Don't worry, weren't you and mruberry planning to fix this? ;) https://github.com/pytorch/pytorch/issues/50953